brave/src/main/java/brave/internal/recorder/OrphanTracker.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright 2013-2020 The OpenZipkin Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package brave.internal.recorder;
+
+import brave.handler.MutableSpan;
+import brave.handler.SpanListener;
+import brave.internal.InternalHandler;
+import brave.internal.Platform;
+import brave.propagation.TraceContext;
+
+// TODO make a weak map of MutableSpan -> Throwable instead of a hacky field in MutableSpan
```

Comment:
PS I plan to shade @raphw's https://github.com/raphw/weak-lock-free/blob/master/src/main/java/com/blogspot/mydailyjava/weaklockfree/WeakConcurrentMap.java into the core jar to support orphan tracker as it is much nicer to use a weak map instead of a hacky field. The code is small (won't make the jar too big), has the same bytecode level as us (Java 6) and stable. We can cite it in the NOTICE file.

If you think this is a bad idea, do scream at some point :D

Reply:
I started to do this, then realized we had hacked this quite a bit in PendingSpans, and also shading adds a lot of complexity and also the class is quite bigger than ours anyway. I cleaned up our variant instead here https://github.com/openzipkin/brave/pull/1152